File: research/ntrip_ingestion/src/nlgcp_ntrip_ingest/sourcetable.py
"""NTRIP source-table retrieval and parsing (safe, bounded).

Only ``STR``/``CAS``/``NET``/``ENDSOURCETABLE`` records are parsed.
Coordinates are provider metadata, never scientifically verified
station coordinates.
"""

from __future__ import annotations

import hashlib
from dataclasses import asdict, dataclass


@dataclass(frozen=True, slots=True)
class StrEntry:
    mountpoint: str
    identifier: str
    format: str
    format_details: str
    carrier: str
    nav_systems: str
    network: str
    country: str
    latitude: str
    longitude: str
    nmea_required: str
    solution: str
    generator: str
    compr_encryp: str
    authentication: str
    fee: str
    bitrate: str
    misc: str = ""

    def as_dict(self) -> dict[str, object]:
        return asdict(self)


@dataclass(frozen=True, slots=True)
class CasEntry:
    host: str
    port: str
    identifier: str
    operator: str = ""
    nmea: str = ""
    country: str = ""
    latitude: str = ""
    longitude: str = ""
    misc: str = ""

    def as_dict(self) -> dict[str, object]:
        return asdict(self)


@dataclass(frozen=True, slots=True)
class NetEntry:
    identifier: str
    operator: str = ""
    authentication: str = ""
    fee: str = ""
    web_net: str = ""
    web_str: str = ""
    web_reg: str = ""
    misc: str = ""

    def as_dict(self) -> dict[str, object]:
        return asdict(self)


@dataclass(slots=True)
class SourceTable:
    streams: list[StrEntry]
    casters: list[CasEntry]
    networks: list[NetEntry]
    terminated: bool
    findings: list[str]
    sha256: str

    def as_dict(self) -> dict[str, object]:
        return {
            "streams": [s.as_dict() for s in self.streams],
            "casters": [c.as_dict() for c in self.casters],
            "networks": [n.as_dict() for n in self.networks],
            "terminated": self.terminated,
            "findings": list(self.findings),
            "sha256": self.sha256,
        }


def _split_fields(line: str) -> list[str]:
    return line.split(";")[1:]


def parse_sourcetable(
    text: str, *, max_bytes: int = 65536, max_entries: int = 10000
) -> SourceTable:
    """Parse a source-table body with explicit size/entry bounds."""
    findings: list[str] = []
    raw = text.encode("utf-8", errors="replace")
    if len(raw) > max_bytes:
        findings.append(
            f"SOURCETABLE_TOO_LARGE bytes={len(raw)} max={max_bytes}"
        )
        raw = raw[:max_bytes]
        text = raw.decode("utf-8", errors="replace")
    table = SourceTable(
        streams=[],
        casters=[],
        networks=[],
        terminated=False,
        findings=findings,
        sha256=hashlib.sha256(raw).hexdigest(),
    )
    total = 0
    for raw_line in text.splitlines():
        line = raw_line.strip()
        if not line:
            continue
        if line == "ENDSOURCETABLE":
            table.terminated = True
            continue
        total += 1
        if total > max_entries:
            table.findings.append(f"ENTRY_LIMIT reached={max_entries}")
            break
        if line.startswith("STR;"):
            fields = _split_fields(line) + [""] * 18
            table.streams.append(
                StrEntry(
                    mountpoint=fields[0],
                    identifier=fields[1],
                    format=fields[2],
                    format_details=fields[3],
                    carrier=fields[4],
                    nav_systems=fields[5],
                    network=fields[6],
                    country=fields[7],
                    latitude=fields[8],
                    longitude=fields[9],
                    nmea_required=fields[10],
                    solution=fields[11],
                    generator=fields[12],
                    compr_encryp=fields[13],
                    authentication=fields[14],
                    fee=fields[15],
                    bitrate=fields[16],
                    misc=fields[17],
                )
            )
        elif line.startswith("CAS;"):
            fields = _split_fields(line) + [""] * 8
            table.casters.append(
                CasEntry(
                    host=fields[0],
                    port=fields[1],
                    identifier=fields[2],
                    operator=fields[3],
                    nmea=fields[4],
                    country=fields[5],
                    latitude=fields[6],
                    longitude=fields[7],
                    misc=fields[8],
                )
            )
        elif line.startswith("NET;"):
            fields = _split_fields(line) + [""] * 7
            table.networks.append(
                NetEntry(
                    identifier=fields[0],
                    operator=fields[1],
                    authentication=fields[2],
                    fee=fields[3],
                    web_net=fields[4],
                    web_str=fields[5],
                    web_reg=fields[6],
                    misc=fields[7],
                )
            )
        else:
            table.findings.append(f"UNKNOWN_RECORD prefix={line[:8]!r}")
    if not table.terminated:
        table.findings.append("MISSING_ENDSOURCETABLE")
    return table


def find_stream(table: SourceTable, mountpoint: str) -> StrEntry | None:
    want = mountpoint.lstrip("/").upper()
    for entry in table.streams:
        if entry.mountpoint.lstrip("/").upper() == want:
            return entry
    return None


def stream_requires_nmea(entry: StrEntry) -> bool:
    return entry.nmea_required.strip() == "1"


def stream_needs_auth(entry: StrEntry) -> bool:
    return entry.authentication.strip().upper() in ("B", "D", "Y", "1")

File: services/gnss-ingestor/internal/ntrip/sourcetable.go
package ntrip

import (
	"strings"
)

// Source-table record kinds: STR (stream), CAS (caster), NET (network).
// Coordinates are provider metadata, never verified station coordinates.

// StrEntry is one parsed STR (stream) record.
type StrEntry struct {
	Mountpoint     string
	Identifier     string
	Format         string
	FormatDetails  string
	Carrier        string
	NavSystems     string
	Network        string
	Country        string
	Latitude       string
	Longitude      string
	NMEARequired   string
	Solution       string
	Generator      string
	ComprEncryp    string
	Authentication string
	Fee            string
	Bitrate        string
	Misc           string
}

// CasEntry is one parsed CAS (caster) record.
type CasEntry struct {
	Host, Port, Identifier, Operator, NMEA, Country, Latitude, Longitude, Misc string
}

// NetEntry is one parsed NET (network) record.
type NetEntry struct {
	Identifier, Operator, Authentication, Fee, WebNet, WebStr, WebReg, Misc string
}

// SourceTable is a parsed, bounded source-table body.
type SourceTable struct {
	Streams    []StrEntry
	Casters    []CasEntry
	Networks   []NetEntry
	Terminated bool
	Findings   []string
	SHA256     string
}

// ParseSourcetable parses a source-table body with explicit size/entry bounds.
func ParseSourcetable(text string, maxBytes, maxEntries int) SourceTable {
	raw := []byte(text)
	table := SourceTable{}
	if len(raw) > maxBytes {
		table.Findings = append(table.Findings, "SOURCETABLE_TOO_LARGE")
		raw = raw[:maxBytes]
		text = string(raw)
	}
	table.SHA256 = SHA256Hex(raw)
	count := 0
	for _, line := range strings.Split(text, "\n") {
		line = strings.TrimRight(strings.TrimSpace(line), "\r")
		if line == "" {
			continue
		}
		if line == "ENDSOURCETABLE" {
			table.Terminated = true
			continue
		}
		count++
		if count > maxEntries {
			table.Findings = append(table.Findings, "ENTRY_LIMIT")
			break
		}
		switch {
		case strings.HasPrefix(line, "STR;"):
			table.Streams = append(table.Streams, parseSTR(line))
		case strings.HasPrefix(line, "CAS;"):
			table.Casters = append(table.Casters, parseCAS(line))
		case strings.HasPrefix(line, "NET;"):
			table.Networks = append(table.Networks, parseNET(line))
		default:
			prefix := line
			if len(prefix) > 8 {
				prefix = prefix[:8]
			}
			table.Findings = append(table.Findings, "UNKNOWN_RECORD prefix="+prefix)
		}
	}
	if !table.Terminated {
		table.Findings = append(table.Findings, "MISSING_ENDSOURCETABLE")
	}
	return table
}

func splitFields(line string, width int) []string {
	fields := strings.Split(line, ";")[1:]
	for len(fields) < width {
		fields = append(fields, "")
	}
	return fields
}

func parseSTR(line string) StrEntry {
	f := splitFields(line, 18)
	return StrEntry{
		Mountpoint: f[0], Identifier: f[1], Format: f[2], FormatDetails: f[3],
		Carrier: f[4], NavSystems: f[5], Network: f[6], Country: f[7],
		Latitude: f[8], Longitude: f[9], NMEARequired: f[10], Solution: f[11],
		Generator: f[12], ComprEncryp: f[13], Authentication: f[14], Fee: f[15],
		Bitrate: f[16], Misc: f[17],
	}
}

func parseCAS(line string) CasEntry {
	f := splitFields(line, 9)
	return CasEntry{
		Host: f[0], Port: f[1], Identifier: f[2], Operator: f[3], NMEA: f[4],
		Country: f[5], Latitude: f[6], Longitude: f[7], Misc: f[8],
	}
}

func parseNET(line string) NetEntry {
	f := splitFields(line, 8)
	return NetEntry{
		Identifier: f[0], Operator: f[1], Authentication: f[2], Fee: f[3],
		WebNet: f[4], WebStr: f[5], WebReg: f[6], Misc: f[7],
	}
}

// FindStream locates a stream by mountpoint (case-insensitive, slash-tolerant).
func FindStream(table SourceTable, mountpoint string) *StrEntry {
	want := strings.ToUpper(strings.TrimPrefix(mountpoint, "/"))
	for i := range table.Streams {
		if strings.ToUpper(strings.TrimPrefix(table.Streams[i].Mountpoint, "/")) == want {
			return &table.Streams[i]
		}
	}
	return nil
}

// StreamRequiresNMEA reports whether the stream demands rover GGA.
func StreamRequiresNMEA(entry *StrEntry) bool {
	return entry != nil && strings.TrimSpace(entry.NMEARequired) == "1"
}

// StreamNeedsAuth reports whether the stream requires authentication.
func StreamNeedsAuth(entry *StrEntry) bool {
	if entry == nil {
		return false
	}
	switch strings.ToUpper(strings.TrimSpace(entry.Authentication)) {
	case "B", "D", "Y", "1":
		return true
	default:
		return false
	}
}
